probe-rs/src/architecture/arm/memory/adi_v5_memory_interface.rs

```diff
@@ -343,28 +341,57 @@ where
         }
     }
 
-    /// Read a 64bit word at `address`.
+    /// Read a block of 64 bit words at `address`.
     ///
+    /// The number of words read is `data.len()`.
     /// The address where the read should be performed at has to be a multiple of 8.
     /// Returns `ArmError::MemoryNotAligned` if this does not hold true.
-    pub fn read_word_64(&mut self, address: u64) -> Result<u64, ArmError> {
+    pub fn read_64(&mut self, mut address: u64, mut data: &mut [u64]) -> Result<(), ArmError> {
+        if data.is_empty() {
+            return Ok(());
+        }
+
         if (address % 8) != 0 {
-            return Err(ArmError::alignment_error(address, 4));
+            return Err(ArmError::alignment_error(address, 8));
         }
 
-        let mut buf = [0u32; 2];
+        // Fall back to 32-bit accesses if 64-bit accesses are not supported.
+        // In both cases the sequence of words we have to read from DRW is the same:
+        // first the least significant word, then the most significant word.
+        let size = match self.ap_information.has_large_data_extension {
+            true => DataSize::U64,
+            false => DataSize::U32,
+        };
+        self.write_csw_register(size)?;
 
-        if !self.ap_information.has_large_data_extension {
-            self.read_32(address, &mut buf)?;
-        } else {
-            let csw = self.build_csw_register(DataSize::U64);
+        while !data.is_empty() {
+            let chunk_size = data.len().min(autoincr_max_bytes(address) / 8);
+
+            tracing::debug!(
+                "Reading chunk with len {} at address {:#08x}",
+                chunk_size,
+                address
+            );
 
-            self.write_csw_register(csw)?;
+            // autoincrement is limited to the 10 lowest bits, so write TAR every time.
             self.write_tar_register(address)?;
+
+            let mut buf = vec![0; chunk_size * 2];
```

Comment:
Could we get rid of this intermediate buffer somehow?

Reply:
can't think of any way other than unsafe transmutes, and then we'd be relying on the endianness of the host which is not great.